Port to LLVM 14

liushuyu opened the pull request:
This pull request port Intel Graphics Compiler to LLVM 14.

I am not sure how to split this pull request into smaller ones. But I am willing to make changes, no matter big or small to this pull request to improve the code quality.

pszymich commented:
Hi @liushuyu,

Thank you very much for this PR.
We are considering merging, but in your patch there are many `#if LLVM_VERSION_MAJOR` blocks. Code that uses such macros should be encapsulated within WrapperLLVM module ([example](https://github.com/intel/intel-graphics-compiler/blob/master/IGC/WrapperLLVM/include/llvmWrapper/IR/DerivedTypes.h)).
Please let us know if you plan on adjusting the PR. Thanks!

liushuyu commented:
> Hi @liushuyu,
> 
> Thank you very much for this PR. We are considering merging, but in your patch there are many `#if LLVM_VERSION_MAJOR` blocks. Code that uses such macros should be encapsulated within WrapperLLVM module ([example](https://github.com/intel/intel-graphics-compiler/blob/master/IGC/WrapperLLVM/include/llvmWrapper/IR/DerivedTypes.h)). Also, make sure not to use tabs and trim EOL whitespace. Please let us know if you plan on adjusting the PR. Thanks!

Thank you for your feedback! I will make the adjustments.

pszymich commented on IGC/AdaptorCommon/LegalizeFunctionSignatures.cpp:
```diff
@@ -548,7 +548,11 @@ void LegalizeFunctionSignatures::FixCallInstruction(Module& M, CallInst* callIns
         {
             // legal argument
             callArgs.push_back(arg);
+#if LLVM_VERSION_MAJOR >= 14
+            ArgAttrVec.push_back(PAL.getParamAttrs(opNum));
+#else
             ArgAttrVec.push_back(PAL.getParamAttributes(opNum));
+#endif
```
Should be wrapped with

```cpp
    llvm::AttributeList getParamAttrs(llvm::AttributeList &AL, unsigned opNum)
    {
#if LLVM_VERSION_MAJOR >= 14
        return AL.getParamAttrs(opNum);
#else
        return AL.getParamAttributes(opNum);
#endif
    }
```

and then use in code:

```suggestion
            ArgAttrVec.push_back(IGCLLVM::getParamAttrs(PAL, opNum));
```

The original llvm code for this class is in `llvm/IR/Attributes.h` so the wrapper should be put in `/IGC/WrapperLLVM/include/llvmWrapper/IR/Attributes.h`.

This is one example. There should be similar wrappers for every situation where `#if LLVM_VERSION_MAJOR` is requried in code, and encapsulated away in the WrapperLLVM module under `IGCLLVM` namespace.

pszymich replied:
Should be wrapped with

```cpp
    llvm::AttributeList getParamAttrs(llvm::AttributeList &AL, unsigned opNum)
    {
#if LLVM_VERSION_MAJOR >= 14
        return AL.getParamAttrs(opNum);
#else
        return AL.getParamAttributes(opNum);
#endif
    }
```

and then use in code:

```suggestion
            ArgAttrVec.push_back(IGCLLVM::getParamAttrs(PAL, opNum));
```

The original llvm code for this class is in `llvm/IR/Attributes.h` so the wrapper should be put in `/IGC/WrapperLLVM/include/llvmWrapper/IR/Attributes.h`.

This is one example. There should be similar wrappers for every situation where `#if LLVM_VERSION_MAJOR` is requried in code, and encapsulated away in the WrapperLLVM module under `IGCLLVM` namespace.

pszymich commented on IGC/AdaptorCommon/LegalizeFunctionSignatures.cpp:
```diff
@@ -581,7 +585,11 @@ void LegalizeFunctionSignatures::FixCallInstruction(Module& M, CallInst* callIns
         // Create the new call instruction
         CallInst* newCallInst = builder.CreateCall(newCalledValue, callArgs);
         newCallInst->setCallingConv(callInst->getCallingConv());
+#if LLVM_VERSION_MAJOR >= 14
+        newCallInst->setAttributes(AttributeList::get(M.getContext(), PAL.getFnAttrs(), PAL.getRetAttrs(), ArgAttrVec));
+#else
         newCallInst->setAttributes(AttributeList::get(M.getContext(), PAL.getFnAttributes(), PAL.getRetAttributes(), ArgAttrVec));
+#endif
```
Same for `llvm::AttributeList::getRetAttributes()`.

pszymich replied:
Same for `getRetAttrs()`.

pszymich commented on IGC/AdaptorCommon/RayTracing/RTBuilder.cpp:
```diff
@@ -3251,30 +3252,42 @@ Value* RTBuilder::createAllocaNumber(const AllocaInst* AI, uint32_t Number)
+    IGCLLVM::AttrBuilder AB { CI->getContext(), Attrs.getAttributes(AttributeList::ReturnIndex)};
     AB.addAlignmentAttr(AlignVal);
     auto AL =
+#if LLVM_VERSION_MAJOR >= 14
+        Attrs.addAttributesAtIndex(CI->getContext(), AttributeList::ReturnIndex, AB);
+#else
         Attrs.addAttributes(CI->getContext(), AttributeList::ReturnIndex, AB);
+#endif
```
Same for `addAttributesAtIndex()`

pszymich commented on IGC/AdaptorOCL/SPIRV/SPIRVReader.cpp:
```diff
@@ -4211,8 +4211,12 @@ SPIRVToLLVM::transFunction(SPIRVFunction *BF) {
   BF->foreachReturnValueAttr([&](SPIRVFuncParamAttrKind Kind){
     if (Kind == FunctionParameterAttributeCount)
       return;
+#if LLVM_VERSION_MAJOR >= 14
+    F->addRetAttr(SPIRSPIRVFuncParamAttrMap::rmap(Kind));
+#else
     F->addAttribute(AttributeList::ReturnIndex,
         SPIRSPIRVFuncParamAttrMap::rmap(Kind));
+#endif
```
Same for `addRetAttr()`.

Etc. for rest of the PR.

pszymich commented on IGC/Compiler/Optimizer/IntDivConstantReduction.cpp:
```diff
@@ -479,8 +483,28 @@ struct IntDivConstantReduction : public FunctionPass
         //
         const int bitSize = dividend->getType()->getIntegerBitWidth();
         //
+#if LLVM_VERSION_MAJOR >= 14
+        SignedDivisionByConstantInfo appxRecip = SignedDivisionByConstantInfo::get(divisor);
+#else
         APInt::ms appxRecip = divisor.magic();
+#endif
         //
+#if LLVM_VERSION_MAJOR >= 14
+        ConstantInt *appxRcp = IGC::getConstantSInt(
+            B, bitSize, appxRecip.Magic.getSExtValue());
+        Value *appxQ =
+            IGC::CreateMulh(F, B, true, dividend, appxRcp);
+        if (divisor.isStrictlyPositive() && appxRecip.Magic.isNegative()) {
+            appxQ = B.CreateAdd(appxQ, dividend, "q_appx");
+        }
+        if (divisor.isNegative() && appxRecip.Magic.isStrictlyPositive()) {
+            appxQ = B.CreateSub(appxQ, dividend, "q_appx");
+        }
+        if (appxRecip.ShiftAmount > 0) {
+            ConstantInt *shift = IGC::getConstantSInt(B, bitSize, appxRecip.ShiftAmount);
+            appxQ = B.CreateAShr(appxQ, shift, "q_appx");
+        }
+#else
```
This seems like more complex example of potential wrapper usage. Having this amount of code under `#if LLVM` isn't desirable.

Maybe there's a way to generalize this under `llvmWrapper/Support/DivisionByConstantInfo.h`?

pszymich commented on IGC/Compiler/Optimizer/IntDivConstantReduction.cpp:
```diff
@@ -520,13 +545,29 @@ struct IntDivConstantReduction : public FunctionPass
         // even divisors can pre-shift the dividend to avoid
         // extra work at the end.
         Value *shiftedDividend = dividend;
+#if LLVM_VERSION_MAJOR >= 14
+        if (appxRecip.IsAdd && !divisor[0]) {
+            unsigned s = divisor.countTrailingZeros();
+            shiftedDividend = B.CreateLShr(shiftedDividend, s);
+            appxRecip = UnsignedDivisonByConstantInfo::get(divisor.lshr(s));
+            IGC_ASSERT_MESSAGE(!appxRecip.IsAdd, "expected to subtract now");
+            IGC_ASSERT_MESSAGE(appxRecip.ShiftAmount < divisor.getBitWidth(), "undefined shift");
+        }
+        //
+        ConstantInt *appxRcp = IGC::getConstantUInt(
+            B, bitSize, appxRecip.Magic.getZExtValue());
+#else
         if (appxRecip.a && !divisor[0]) {
             unsigned s = divisor.countTrailingZeros();
             shiftedDividend = B.CreateLShr(shiftedDividend, s);
```
There is basically the same code under `#if` and `#else`. We cannot accept this. If `ShiftAmount` & `IsAdd` are the only differences we need to put this is the wrapper somehow.

pszymich commented on IGC/Compiler/Optimizer/IntDivConstantReduction.cpp:
```diff
@@ -537,9 +578,20 @@ struct IntDivConstantReduction : public FunctionPass
+#endif
         Value *appxQ =
             IGC::CreateMulh(F, B, false, shiftedDividend, appxRcp);
         //
+#if LLVM_VERSION_MAJOR >= 14
+        if (!appxRecip.IsAdd) {
+            appxQ = B.CreateLShr(appxQ, appxRecip.ShiftAmount, "q_appx");
+        } else {
+            Value *fixup = B.CreateSub(dividend, appxQ, "q_appx");
+            fixup = B.CreateLShr(fixup, 1);
+            appxQ = B.CreateAdd(fixup, appxQ, "q_appx");
+            appxQ = B.CreateLShr(appxQ, appxRecip.ShiftAmount - 1, "q_appx");
+        }
+#else
         if (!appxRecip.a) {
             appxQ = B.CreateLShr(appxQ, appxRecip.s, "q_appx");
```
Same here

pszymich commented on IGC/Compiler/Optimizer/OpenCLPasses/AlignmentAnalysis/AlignmentAnalysis.cpp:
```diff
@@ -165,7 +165,11 @@ unsigned int AlignmentAnalysis::visitSelectInst(SelectInst& I)
 
 unsigned int AlignmentAnalysis::visitPHINode(PHINode& I)
 {
+#if LLVM_VERSION_MAJOR >= 14
+    uint64_t newAlign = Value::MaximumAlignment;
+#else
     unsigned int newAlign = Value::MaximumAlignment;
+#endif
```
You can use `auto` here, however it's good to make a quick code check and make sure there are no obvious problems. Somethimes `auto` can behave unexpectedly.

pszymich commented on IGC/Compiler/Optimizer/OpenCLPasses/AlignmentAnalysis/AlignmentAnalysis.cpp:
```diff
@@ -440,9 +444,13 @@ void AlignmentAnalysis::SetInstAlignment(MemSetInst& I)
 #if LLVM_VERSION_MAJOR == 4
     unsigned alignment = iSTD::Max(I.getAlignment(), getAlignValue(I.getRawDest()));
     I.setAlignment(ConstantInt::get(Type::getInt32Ty(I.getContext()), alignment));
-#elif LLVM_VERSION_MAJOR >= 7
+#elif LLVM_VERSION_MAJOR >= 7 && LLVM_VERSION_MAJOR < 14
     unsigned alignment = iSTD::Max(I.getDestAlignment(), getAlignValue(I.getRawDest()));
     I.setDestAlignment(alignment);
+#elif LLVM_VERSION_MAJOR >= 14
+    uint64_t alignment_value = iSTD::Max(I.getDestAlign()->value(), getAlignValue(I.getRawDest()));
+    llvm::Align alignment = llvm::Align(alignment_value);
+    I.setDestAlignment(alignment);
 #endif
```
You can drop LLVM 4 support here and leave:

```cpp
#if LLVM_VERSION_MAJOR >= 14
    uint64_t alignment_value = iSTD::Max(I.getDestAlign()->value(), getAlignValue(I.getRawDest()));
    llvm::Align alignment = llvm::Align(alignment_value);
    I.setDestAlignment(alignment);
#else
    unsigned alignment = iSTD::Max(I.getDestAlignment(), getAlignValue(I.getRawDest()));
    I.setDestAlignment(alignment);
#endif
```

From brief analysis: this is a subject to refactor to remove `#if` guard using this [`getAlign()`](https://github.com/intel/intel-graphics-compiler/blob/master/IGC/WrapperLLVM/include/llvmWrapper/Support/Alignment.h#L73) wrapper and adding whatever is necessary for LLVM 14, from the looks of it wrapper on `I.getDestAlign()`/`I.getAlignment()`.

`llvmWrapper/Support/Alignment.h` provides tools to deal with alignment problems between LLVM versions.

liushuyu replied:
> From brief analysis: this is a subject to refactor to remove `#if` guard using this [`getAlign()`](https://github.com/intel/intel-graphics-compiler/blob/master/IGC/WrapperLLVM/include/llvmWrapper/Support/Alignment.h#L73) wrapper and adding whatever is necessary for LLVM 14, from the looks of it wrapper on `I.getDestAlign()`/`I.getAlignment()`.

I think I will leave the refactor to some other people since I am not very familiar with this codebase.

pszymich commented on IGC/Compiler/Optimizer/OpenCLPasses/AlignmentAnalysis/AlignmentAnalysis.cpp:
```diff
@@ -472,7 +480,11 @@ void AlignmentAnalysis::SetInstAlignment(MemMoveInst& I)
 #endif
 }
 
+#if LLVM_VERSION_MAJOR >= 14
+uint64_t AlignmentAnalysis::getAlignValue(Value* V) const
+#else
 unsigned int AlignmentAnalysis::getAlignValue(Value* V) const
+#endif
```
Same remarks about `auto`.

pszymich commented on IGC/Compiler/Optimizer/OpenCLPasses/AlignmentAnalysis/AlignmentAnalysis.cpp:
```diff
@@ -541,6 +553,17 @@ unsigned int AlignmentAnalysis::getAlignValue(Value* V) const
     return MinimumAlignment;
 }
 
+#if LLVM_VERSION_MAJOR >= 14
+uint64_t AlignmentAnalysis::getConstantAlignment(uint64_t C) const
+{
+    if (!C)
+    {
+        return Value::MaximumAlignment;
+    }
+
+    return iSTD::Min(Value::MaximumAlignment, 1UL << llvm::countTrailingZeros(C));
+}
+#else
 unsigned int AlignmentAnalysis::getConstantAlignment(uint64_t C) const
 {
     if (!C)
```
This doesn't seem necessary, if the only difference is the retval.

pszymich commented on IGC/DebugInfo/StreamEmitter.cpp:
```diff
@@ -322,7 +322,11 @@ class VISAAsmBackend : public MCAsmBackend {
   }
 #endif
 
+#if LLVM_VERSION_MAJOR >= 14
+  bool writeNopData(raw_ostream &OS, uint64_t Count, const MCSubtargetInfo *STI) const override {
+#else
   bool writeNopData(raw_ostream &OS, uint64_t Count) const override {
+#endif
```
Easy wrapper for this, just proxy `writeNopData` call like this:

```cpp
    T f(a, b)
    {
#if LLVM14
        return g(a, b);
#else
        return g(a);
#endif
    }
```

liushuyu replied:
> Easy wrapper for this, just proxy `writeNopData` call like this:

Could you elaborate on how to do this when the function is `override`'ing a virtual function? I can't see how I can use a wrapper for this.

pszymich commented on IGC/VectorCompiler/lib/GenXCodeGen/GenXCoalescing.cpp:
```diff
@@ -203,6 +202,8 @@ SPDX-License-Identifier: MIT
 #include <map>
 #include <vector>
 
+#define DEBUG_TYPE "GENX_COALESCING"
+
```
What's this for? I see reference here: https://llvm.org/docs/ProgrammersManual.html, but I'm not sure why it's needed here for LLVM 14 compatibility.

liushuyu replied:
> What's this for? I see reference here: https://llvm.org/docs/ProgrammersManual.html, but I'm not sure why it's needed here for LLVM 14 compatibility.

There are some headers in the `#include` that has `#undef DEBUG_TYPE`. They were added in LLVM 14 for some reason.

pszymich commented on IGC/VectorCompiler/lib/GenXCodeGen/GenXVisa.h:
```diff
@@ -30,6 +30,12 @@ SPDX-License-Identifier: MIT
 #include <string>
 #include <vector>
 
+#if LLVM_VERSION_MAJOR >= 14
+typedef uint64_t alignment_t;
+#else
+typedef unsigned alignment_t;
+#endif
```
That's provided within `llvmWrapper/Support/Alignment.h`

pszymich commented on IGC/WrapperLLVM/include/llvmWrapper/IR/Attributes.h:
```diff
@@ -0,0 +1,58 @@
+#endif
+  }
+
+/// Return the attribute object that exists at the given index.
```
There's an indent missing here. Please use 4 space indents for the file.

pszymich commented on IGC/WrapperLLVM/include/llvmWrapper/IR/ConstantFolder.h:
```diff
@@ -31,11 +31,85 @@ namespace IGCLLVM
+            llvm::ArrayRef<llvm::Constant*> IdxList) const override {
+            return m_baseConstantFolder.CreateInBoundsGetElementPtr(Ty, C, IdxList);
+        }
+
+        inline llvm::Constant* CreateInBoundsGetElementPtr(llvm::Type* Ty, llvm::Constant* C,
+            llvm::Constant* Idx) const override {
+            return m_baseConstantFolder.CreateInBoundsGetElementPtr(Ty, C, Idx);
+        }
+
+        inline llvm::Constant* CreateInBoundsGetElementPtr(llvm::Type* Ty, llvm::Constant* C,
+            llvm::ArrayRef<llvm::Value*> IdxList) const override {
+            return m_baseConstantFolder.CreateInBoundsGetElementPtr(Ty, C, IdxList);
+        }
+#else
+        inline llvm::Value* FoldAdd(llvm::Value* LHS, llvm::Value* RHS, bool HasNUW = false,
+            bool HasNSW = false) const override {
+            return m_baseConstantFolder.FoldAdd(LHS, RHS, HasNUW, HasNSW);
+        }
+
+        inline llvm::Value* FoldAnd(llvm::Value* LHS, llvm::Value* RHS) const override {
+            return m_baseConstantFolder.FoldAnd(LHS, RHS);
+        }
+
+        inline llvm::Value* FoldOr(llvm::Value* LHS, llvm::Value* RHS) const override {
+            return m_baseConstantFolder.FoldOr(LHS, RHS);
+        }
+
+        inline llvm::Value* FoldICmp(llvm::CmpInst::Predicate P, llvm::Value* LHS, llvm::Value* RHS) const override {
+            return m_baseConstantFolder.FoldICmp(P, LHS, RHS);
+        }
+
+        inline llvm::Value* FoldSelect(llvm::Value* C, llvm::Value* True, llvm::Value* False) const override {
+            return m_baseConstantFolder.FoldSelect(C, True, False);
+        }
+
+        inline llvm::Value* FoldGEP(llvm::Type* Ty, llvm::Value* Ptr, llvm::ArrayRef<llvm::Value*> IdxList,
+            bool IsInBounds = false) const override {
+            return m_baseConstantFolder.FoldGEP(Ty, Ptr, IdxList, IsInBounds);
+        }
+#endif
```
Can you summarize what's going on here?

liushuyu replied:
> Can you summarize what's going on here?

In LLVM 14, they have rewritten the whole `llvm::IRBuilderFolder` class. So the previous overloading no longer works due to renamed member functions.

pszymich replied:
The way the code is right now eg. `CreateOr()` wouldn't be available in LLVM >= 14 would it?
The wrapper idea is to have one symbol in core IGC code that proxies implementation depending on LLVM version. If `Create*` functions are for LLVM < 14, and `Fold*` for LLVM >= 14 that misses the point because you would still need `#if LLVM` guards in code depending on what LLVM you are linking against.

Does `CreateAdd()` work the same as `FoldAdd()`? Is so the wrapper may be written like this:

```cpp
T fold*(...)
{
#if LLVM 14
  ret llvm::fold*(...)
#else
  ret llvm::create*(...)
#endif
}
```

liushuyu replied:
Upon close inspection, isn't this class actually a wrapper itself to just proxy the function calls to the wrapped base class?

> The wrapper idea is to have one symbol in core IGC code that proxies implementation depending on LLVM version. If `Create*` functions are for LLVM < 14, and `Fold*` for LLVM >= 14 that misses the point because you would still need `#if LLVM` guards in code depending on what LLVM you are linking against.

It does make sense to do this because I think the purpose of this class is to inherit a base class to avoid the `final` modifier?

pszymich replied:
Yes, that's correct. Since some recent LLVM version the class became `final` and forced us to create this folder "proxy".

I see in the LLVM differential revisions, such as here: https://reviews.llvm.org/D117431, or here: https://reviews.llvm.org/D116968 that there's not much difference between certain `Create` and `Fold` versions, aside from additional type check. Some cases tough, like this change for GEPs: https://reviews.llvm.org/D117039 is much more nuanced and may warrant a separate PR just to accomodate it.

There are good and worse ways of doing this. I understand this gets complicated but the folder wrapper should ideally migrate to `Fold`s and implement these such that on LLVM < 14 we get `Create`s' functionality. This will cause more refactoring since every previous `Create` user will have to be changed to `Fold`, however I'm not seeing a better way to do this. Keeping both `Create`s and `Fold`s does not seem sensible, since LLVM didn't either, instead of deprecating. Having `Create`s or `Fold`s depending on LLVM version breaks API compatibility within IGC.

liushuyu replied:
> There are good and worse ways of doing this. I understand this gets complicated but the folder wrapper should ideally migrate to `Fold`s and implement these such that on LLVM < 14 we get `Create`s' functionality. This will cause more refactoring since every previous `Create` user will have to be changed to `Fold`, however I'm not seeing a better way to do this. Keeping both `Create`s and `Fold`s does not seem sensible, since LLVM didn't either, instead of deprecating. Having `Create`s or `Fold`s depending on LLVM version breaks API compatibility within IGC.

It seems strange because it did not give me any compile errors when I changed the code like this. Probably there is currently (as of writing) nothing is using `#ifdef`ed APIs.

I am not sure what I should do here. Should I add non-`override` variants of the newer APIs to the older LLVM version of the wrapper? Since removing `override` functions will cause the class to become a virtual class.

pszymich replied:
> nothing is using #ifdefed APIs

The IGC repository is also a part of our proprietary drivers, so we cannot assume this.

> Should I add non-override variants of the newer APIs to the older LLVM version of the wrapper?

Ideally yes, assumming corresponding `Fold` is a renamed `Create`. Then change users of these functions to use `Fold`. The goal is to have one symbol to use regardless of underlying LLVM version.
This should work as expected:
```cpp
T Fold() const
#if LLVM 14
override {
    return m_base.Fold*();
}
#else
{
    return m_base.Create*();
}
#endif
```
https://godbolt.org/z/hPT8GTf7d

liushuyu replied:
Done.

pszymich reviewed (changes requested):
Hi, there are still some problems, mainly with `#if LLVM_VERSION_MAJOR` code outside of LLVM wrapper. Please refer to the comments. Thanks!

liushuyu commented:
> Hi, there are still some problems, mainly with `#if LLVM_VERSION_MAJOR` code outside of LLVM wrapper. Please refer to the comments. Thanks!

Thank you for your detailed review! I will make the changes outlined in your review.

pszymich commented on IGC/WrapperLLVM/include/llvmWrapper/IR/IRBuilder.h:
```diff
@@ -301,6 +301,14 @@ namespace IGCLLVM
 
 #endif
 
+#if LLVM_VERSION_MAJOR >= 14
+        Value* CreatePtrDiff(Value *LHS, Value *RHS, const Twine &Name = "") {
+          auto *PtrTy = cast<PointerType>(LHS->getType());
+          Type *Ty = PtrTy->getElementType();
+          return llvm::IRBuilder<T, InserterTyDef()>::CreatePtrDiff(Ty, LHS, RHS, Name);
+        }
+#endif
```
Does `CreatePtrDiff` exist in LLVM < 14? Will this be available on LLVM < 14?

liushuyu replied:
> Does `CreatePtrDiff` exist in LLVM < 14? Will this be available on LLVM < 14?

Upon investigation, it seems like this API exists since LLVM 3.x era. So should I just remove the `#ifdef` altogether?

pszymich replied:
Sure, thanks!

liushuyu replied:
I realized I have to actually add this `#ifdef` back since the reason I added this in the first place was because the API changed in LLVM 14 not because the API was not available.

pszymich commented on IGC/Compiler/Optimizer/OpenCLPasses/RayTracing/ResolveOCLRaytracingBuiltins.cpp:
```diff
@@ -134,7 +134,7 @@ void* __builtin_IB_intel_get_rt_stack( rtglobals_t rt_dispatch_globals );
 The RT Stack address is computed as:
     syncStackSize = sizeof(HitInfo)*2 + (sizeof(Ray) + sizeof(TravStack))*RTDispatchGlobals.maxBVHLevels;
-    syncBase = RTDispatchGlobals.rtMemBasePtr � (DSSID * NUM_SIMD_LANES_PER_DSS + StackID + 1)*syncStackSize;
+    syncBase = RTDispatchGlobals.rtMemBasePtr � (DSSID * NUM_SIMD_LANES_PER_DSS + StackID + 1)*syncStackSize;
```
What's this? 🙂

pszymich commented on IGC/DebugInfo/StreamEmitter.cpp:
```diff
@@ -417,7 +426,11 @@ StreamEmitter::StreamEmitter(raw_pwrite_stream &outStream,
                                     std::move(pObjectWriter),
                                     std::move(pCodeEmitter), isRelaxAll);
 
+#if LLVM_VERSION_MAJOR >= 14
+  m_pMCStreamer->initSections(isNoExecStack, *m_pContext->getSubtargetInfo());
+#else
   m_pMCStreamer->InitSections(isNoExecStack);
+#endif
```
Can we have:
```cpp
IGCLLVM::initSections(&m_pMCStreamer, isNoExecStack, *m_pContext->getSubtargetInfo());
```
added in the wrapper?

pszymich commented on IGC/LLVM3DBuilder/BuiltinsFrontendDefinitions.hpp:
```diff
@@ -3736,7 +3736,10 @@ inline llvm::Value* LLVM3DBuilder<preserveNames, T, Inserter>::create_wavePrefix
+#if LLVM_VERSION_MAJOR >= 14
+        CI->setOnlyAccessesInaccessibleMemory();
+        CI->removeAttributeAtIndex(llvm::AttributeList::FunctionIndex, llvm::Attribute::ReadNone);
+#elif LLVM_VERSION_MAJOR >= 7
```
`#elif LLVM >= 7` can be used always, and `#else // LLVM < 7` dropped.

pszymich commented:
Hi @liushuyu,

The changes look good to me. I started the verification process your change need to pass to be merged. I'll keep you updated here.

Thanks!

liushuyu commented:
> [...] Please also check your PR diff for EOL whitespace, as well as possible tab characters.

Normally I would use `clang-format` to remove EOL whitespace and normalize whitespace characters, but I can't seem to find a suitable preset for this repository.

I am not sure if you have a `clang-format` ruleset available, but this would come in handy in situations like this.

pszymich commented:
We don't enforce `clang-format` everywhere and there's no unified ruleset, sorry.

Regarding verification, from Linux Ubuntu LLVM 13 config:
```logs
20:19:35,261 WARN  - /opt/src/igc/IGC/VectorCompiler/lib/GenXCodeGen/TargetInfo/GenXTargetInfo.cpp:11:10: fatal error: llvmWrapper/Support/TargetRegistry.h: No such file or directory
20:19:35,261 WARN  -    11 | #include "llvmWrapper/Support/TargetRegistry.h"
20:19:35,261 WARN  -       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
20:19:35,261 WARN  - compilation terminated.
```
Here's the place to add this as included file: https://github.com/intel/intel-graphics-compiler/blob/master/IGC/WrapperLLVM/CMakeLists.txt#L44

liushuyu commented:
> Here's the place to add this as included file: https://github.com/intel/intel-graphics-compiler/blob/master/IGC/WrapperLLVM/CMakeLists.txt#L44

Should be fixed now.

pszymich commented:
From Linux Ubuntu LLVM 11 validation build:

```logs
14:45:10,409 WARN  - /opt/src/igc/IGC/DebugInfo/StreamEmitter.cpp:432:37: error: 'class llvm::MCContext' has no member named 'getSubtargetInfo'; did you mean 'getSubtargetCopy'?
14:45:10,409 WARN  -                          m_pContext->getSubtargetInfo());
14:45:10,409 WARN  -                                      ^~~~~~~~~~~~~~~~
14:45:10,409 WARN  -                                      getSubtargetCopy
14:45:10,409 WARN  - make[2]: *** [IGC/DebugInfo/CMakeFiles/GenXDebugInfo.dir/StreamEmitter.cpp.o] Error 1
```

liushuyu commented:
> From Linux Ubuntu LLVM 11 validation build:
> 
> ```
> 14:45:10,409 WARN  - /opt/src/igc/IGC/DebugInfo/StreamEmitter.cpp:432:37: error: 'class llvm::MCContext' has no member named 'getSubtargetInfo'; did you mean 'getSubtargetCopy'?
> 14:45:10,409 WARN  -                          m_pContext->getSubtargetInfo());
> 14:45:10,409 WARN  -                                      ^~~~~~~~~~~~~~~~
> 14:45:10,409 WARN  -                                      getSubtargetCopy
> 14:45:10,409 WARN  - make[2]: *** [IGC/DebugInfo/CMakeFiles/GenXDebugInfo.dir/StreamEmitter.cpp.o] Error 1
> ```
> 

I have changed `IGCLLVM::initSections` to accept a `llvm::MCContext` instead to avoid the `getSubtargetInfo` call (which does not exist in LLVM 11).

> From Linux Ubuntu LLVM 13 validation build:
> 
> ```
> 14:51:06,619 WARN  - In file included from /usr/include/c++/9/cassert:44,
> 14:51:06,619 WARN  -                  from /opt/src/igc/IGC/Probe/Assertion.h:16,
> 14:51:06,619 WARN  -                  from /opt/src/igc/IGC/Compiler/Optimizer/OpenCLPasses/ImageFuncs/ResolveSampledImageBuiltins.cpp:15:
> 14:51:06,619 WARN  - /opt/src/igc/IGC/Compiler/Optimizer/OpenCLPasses/ImageFuncs/ResolveSampledImageBuiltins.cpp: In member function 'llvm::Value* IGC::ResolveSampledImageBuiltins::lowerGetImage(llvm::CallInst&)':
> 14:51:06,619 WARN  - /opt/src/igc/IGC/Compiler/Optimizer/OpenCLPasses/ImageFuncs/ResolveSampledImageBuiltins.cpp:99:16: error: 'IGCLLVM' has not been declared
> 14:51:06,619 WARN  -    99 |     IGC_ASSERT(IGCLLVM::getNumArgOperands(&CI) == 1);
> 14:51:06,619 WARN  -       |                ^~~~~~~
> ```

Fixed by adding an include.

I am not sure why I can't reproduce those issues locally with LLVM 13.0.1. Probably because I have a newer CMake (cmake version 3.23.1)?

pszymich commented:
Latest fixes were enough for initial verfication builds to pass. Thank you!
We are now running extended test coverage on your change. Initial results revealed some problems. I'll be fixing them personally keep you up to date with the current status.

pszymich commented on IGC/WrapperLLVM/include/llvmWrapper/IR/ConstantFolder.h:
```diff
@@ -11,9 +11,11 @@ SPDX-License-Identifier: MIT
 
 namespace IGCLLVM
 {
+    using namespace llvm;
```
Please remove the `using` directive from here and use explicit scoping throughout WrapperLLVM module.

liushuyu replied:
Done. And rebased.

liushuyu commented:
Should I rebase this pull request? What's the status of those changes?

frantisekz commented:
@liushuyu 

I am getting these errors while compiling master with this rebased (with llvm/clang 14):

~~~~
In file included from /builddir/build/BUILD/intel-graphics-compiler-igc-1.0.11279/IGC/VectorCompiler/lib/GenXOpts/CMTrans/CMABI.cpp:26:
/builddir/build/BUILD/intel-graphics-compiler-igc-1.0.11279/IGC/WrapperLLVM/include/llvmWrapper/IR/Attributes.h:20:42: error: use of undeclared identifier 'AttributeList'; did you mean 'llvm::AttributeList'?
        return Attrs.getAttributeAtIndex(AttributeList::FunctionIndex, Kind);
                                         ^~~~~~~~~~~~~
                                         llvm::AttributeList
/usr/include/llvm/IR/Intrinsics.h:31:7: note: 'llvm::AttributeList' declared here
class AttributeList;
      ^
In file included from /builddir/build/BUILD/intel-graphics-compiler-igc-1.0.11279/IGC/VectorCompiler/lib/GenXOpts/CMTrans/CMABI.cpp:26:
/builddir/build/BUILD/intel-graphics-compiler-igc-1.0.11279/IGC/WrapperLLVM/include/llvmWrapper/IR/Attributes.h:29:42: error: use of undeclared identifier 'AttributeList'; did you mean 'llvm::AttributeList'?
        return Attrs.getAttributeAtIndex(AttributeList::FunctionIndex, Kind);
                                         ^~~~~~~~~~~~~
                                         llvm::AttributeList
/usr/include/llvm/IR/Intrinsics.h:31:7: note: 'llvm::AttributeList' declared here
class AttributeList;
      ^
In file included from /builddir/build/BUILD/intel-graphics-compiler-igc-1.0.11279/IGC/VectorCompiler/lib/GenXOpts/CMTrans/CMABI.cpp:26:
/builddir/build/BUILD/intel-graphics-compiler-igc-1.0.11279/IGC/WrapperLLVM/include/llvmWrapper/IR/Attributes.h:62:9: error: constructor for 'IGCLLVM::AttrBuilder' must explicitly initialize the base class 'llvm::AttrBuilder' which does not have a default constructor
        AttrBuilder(llvm::LLVMContext &Ctx) {
        ^
/usr/include/llvm/IR/Attributes.h:996:7: note: 'llvm::AttrBuilder' declared here
class AttrBuilder {
      ^
In file included from /builddir/build/BUILD/intel-graphics-compiler-igc-1.0.11279/IGC/VectorCompiler/lib/GenXOpts/CMTrans/CMABI.cpp:26:
/builddir/build/BUILD/intel-graphics-compiler-igc-1.0.11279/IGC/WrapperLLVM/include/llvmWrapper/IR/Attributes.h:64:31: error: redefinition of 'Ctx' with a different type: 'llvm::AttrBuilder' vs 'llvm::LLVMContext &'
            llvm::AttrBuilder(Ctx);
                              ^
/builddir/build/BUILD/intel-graphics-compiler-igc-1.0.11279/IGC/WrapperLLVM/include/llvmWrapper/IR/Attributes.h:62:40: note: previous definition is here
        AttrBuilder(llvm::LLVMContext &Ctx) {
                                       ^
/builddir/build/BUILD/intel-graphics-compiler-igc-1.0.11279/IGC/WrapperLLVM/include/llvmWrapper/IR/Attributes.h:71:9: error: constructor for 'IGCLLVM::AttrBuilder' must explicitly initialize the base class 'llvm::AttrBuilder' which does not have a default constructor
        AttrBuilder(llvm::LLVMContext &Ctx, llvm::AttributeSet AS) {
        ^
/usr/include/llvm/IR/Attributes.h:996:7: note: 'llvm::AttrBuilder' declared here
class AttrBuilder {
      ^
5 errors generated.
~~~~

Do you have any idea what could be the issue?

zboszor commented:
@frantisekz 
> I am getting these errors while compiling master with this rebased (with llvm/clang 14):
> ... 
> Do you have any idea what could be the issue?

The errors are spelled out very verbosely. "Something is undeclared. It's declared in this header: ..."
Can you try this patch? It should at least fix some of the errors. I don't know if it fixes the `redefinition of 'Ctx'` error.
```
diff --git a/IGC/VectorCompiler/lib/GenXOpts/CMTrans/CMABI.cpp b/IGC/VectorCompiler/lib/GenXOpts/CMTrans/CMABI.cpp
index d84988ce3..d76d48a75 100644
--- a/IGC/VectorCompiler/lib/GenXOpts/CMTrans/CMABI.cpp
+++ b/IGC/VectorCompiler/lib/GenXOpts/CMTrans/CMABI.cpp
@@ -26,6 +26,7 @@ SPDX-License-Identifier: MIT
 #include "llvmWrapper/Analysis/CallGraph.h"
 #include "llvmWrapper/IR/DerivedTypes.h"
 #include "llvmWrapper/IR/Instructions.h"
+#include "llvmWrapper/IR/Intrinsics.h"
 #include "llvmWrapper/Support/Alignment.h"
 
 #include "Probe/Assertion.h"
@@ -54,6 +55,7 @@ SPDX-License-Identifier: MIT
 #include "llvm/Analysis/PostDominators.h"
 #include "llvm/GenXIntrinsics/GenXIntrinsics.h"
 #include "llvm/GenXIntrinsics/GenXMetadata.h"
+#include "llvm/IR/Attributes.h"
 #include "llvm/IR/CFG.h"
 #include "llvm/IR/DebugInfo.h"
 #include "llvm/IR/DiagnosticInfo.h"
```

pszymich commented:
> Should I rebase this pull request? What's the status of those changes?

As said here https://github.com/intel/intel-graphics-compiler/pull/242#issuecomment-1132783827 we ran into validation issues when preparing to merge the PR. I'm OoO this week, please expect more information next week.

alucryd commented:
Looks like it might need some more changes for LLVM 14.0.4?

https://paste.xinu.at/Z0JQq7RZLcJeRYIW/

liushuyu commented:
> Looks like it might need some more changes for LLVM 14.0.4?

Thanks to bring that to my attention. The reason for the breakage is two-fold: (1) LLVM 14.0.4 has marked some APIs for deprecation and IGC enabled `-Werror`; (2) newly added changes (relative to this pull request) in IGC did not take LLVM 14 into account.

Now I have fixed the issues in the newly rebased commit.

> 
> https://paste.xinu.at/Z0JQq7RZLcJeRYIW/

From the looks of it, you seem to be packaging for Arch Linux. In which case, you might want to format the patch from another branch in my repository, which is based on a tagged version (version `1.0.11279`): https://github.com/liushuyu/intel-graphics-compiler/tree/llvm-14.

pszymich commented:
Hi, the PR has been validated for all configurations. We're now running extended testing. Thank you for patience.

pszymich commented:
Hi @liushuyu, your change has been merged and you can see your commit here: 03da8989d0d4ca67d21a2b2b86697e97e3561306
We will follow up with setting up LLVM 14 builds in our public CI.
Thank you for the contribution!
